Kernel/Graphics/VirtIOGPU/GraphicsAdapter.h

```diff
@@ -37,7 +95,49 @@ class GraphicsAdapter final
     virtual bool try_to_set_resolution(size_t, size_t, size_t) override { return false; }
     virtual bool set_y_offset(size_t, size_t) override { return false; }
 
-    RefPtr<GPU> m_gpu_device;
     bool m_created_framebuffer_devices { false };
+
+    struct Scanout {
+        RefPtr<Graphics::VirtIOGPU::FramebufferDevice> framebuffer;
+        RefPtr<Console> console;
+        Protocol::DisplayInfoResponse::Display display_info {};
+    };
+
+    virtual bool handle_device_config_change() override;
+    virtual void handle_queue_update(u16 queue_index) override;
+    u32 get_pending_events();
+    void clear_pending_events(u32 event_bitmask);
+
+    auto& operation_lock() { return m_operation_lock; }
+    ResourceID allocate_resource_id();
+
+    PhysicalAddress start_of_scratch_space() const { return m_scratch_space->physical_page(0)->paddr(); }
+    AK::BinaryBufferWriter create_scratchspace_writer()
+    {
+        return { Bytes(m_scratch_space->vaddr().as_ptr(), m_scratch_space->size()) };
+    }
+    void synchronous_virtio_gpu_command(PhysicalAddress buffer_start, size_t request_size, size_t response_size);
+    void populate_virtio_gpu_request_header(Protocol::ControlHeader& header, Protocol::CommandType ctrl_type, u32 flags = 0);
+
+    void query_display_information();
+    ResourceID create_2d_resource(Protocol::Rect rect);
+    void delete_resource(ResourceID resource_id);
+    void ensure_backing_storage(Memory::Region const& region, size_t buffer_offset, size_t buffer_length, ResourceID resource_id);
+    void detach_backing_storage(ResourceID resource_id);
+    void set_scanout_resource(ScanoutID scanout, ResourceID resource_id, Protocol::Rect rect);
+    void transfer_framebuffer_data_to_host(ScanoutID scanout, Protocol::Rect const& rect, ResourceID resource_id);
+    void flush_displayed_image(Protocol::Rect const& dirty_rect, ResourceID resource_id);
```

Comment:
While you are refactoring this, could you change the order of the parameters here? It's a bit ugly that everywhere else uses ResourceID/ScanoutID as first parameter, but this one doesn't.

Reply:
I agree this doesn't look too much neat, so I'll fix it in a separate commit.